File: leetcode/merge-intervals.go
/*
- Problem: https://leetcode.com/problems/merge-intervals/
*/

func merge(intervals [][]int) [][]int {
    
    // First, sorting the 2d array based on the first index of each sub array
    sort.Slice(intervals, func(i, j int) bool { return intervals[i][0] < intervals[j][0] })
    
    var result [][]int = [][]int{}
  
    // Initializing first merging interval
    var tmp []int = intervals[0]
    
    for _, intvl := range intervals {
        // If the last value of merging interval is more than the first value of the current interval, then these 2 needs to be merged. 
        // Or here breaks the current merging interval and starts a new one.
        if tmp[1] < intvl[0] {
            result = append(result, tmp)
            tmp = intvl
        } else {
            if tmp[1] < intvl[1] {
                tmp[1] = intvl[1]
                continue
            }
        }
    }
    
    result = append(result, tmp)
    return result   
}

File: leetcode/java/merge-intervals.java
/*
- Problem: https://leetcode.com/problems/merge-intervals/
*/

class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        List<int[]> result = new ArrayList<int[]>();

        int[] mergingArray = intervals[0];

        for (int i = 1; i < intervals.length; i++) {
            if (mergingArray[1] < intervals[i][0]) {
                // Breaking the current merge and starting a new one.
                result.add(mergingArray);
                mergingArray = intervals[i];
            } else {
                // Increasing the merging interval length if required.
                mergingArray[1] = intervals[i][1] > mergingArray[1] ? intervals[i][1] : mergingArray[1];
                continue;
            }
        }

        result.add(mergingArray);

        return result.stream().toArray(int[][]::new);
    }
}
